implementations/rust/ockam/ockam_identity/src/channel/decryptor_worker.rs

```diff
@@ -670,31 +627,57 @@ impl Worker for DecryptorWorker {
     ) -> Result<()> {
         let msg_addr = msg.msg_addr();
 
-        match self.state()? {
-            State::KeyExchange => {
-                if msg_addr == self.addresses.decryptor_remote {
-                    self.handle_key_exchange_msg(ctx, msg).await?;
-                } else {
-                    return Err(IdentityError::UnknownChannelMsgDestination.into());
-                }
+        // since once initialized, the state can't change anymore we just borrow a mutable,
+        // doing so we avoid extra copies (in case the compiler don't optimize them away),
+        // and we make sure the state remain initialized even in the case of an error
+        if let State::Initialized(state) = self
+            .state
+            .as_mut()
+            .ok_or_else(|| IdentityError::InvalidSecureChannelInternalState)?
+        {
+            if msg_addr == state.addresses.decryptor_remote {
+                state.handle_decrypt(ctx, msg).await?;
+            } else if msg_addr == state.addresses.decryptor_api {
+                state.handle_decrypt_api(ctx, msg).await?;
+            } else {
+                return Err(IdentityError::UnknownChannelMsgDestination.into());
             }
-            State::ExchangeIdentity => {
-                if msg_addr == self.addresses.decryptor_remote {
-                    self.handle_exchange_identity(ctx, Some(msg)).await?;
+
+            return Ok(());
+        }
+
+        let state = self
+            .state
+            .take()
+            .ok_or_else(|| IdentityError::InvalidSecureChannelInternalState)?;
+
+        // if any error occurs during key and identity exchange the state will become
+        // invalid, this enforces atomicity reducing the possibility of attacks
```

Comment:
Shouldn't we close the context and kill the worker in that case?

Reply:
The second `handle_message()` after an error may or may not happen, when you are debugging, you want a consistent behavior: either the worker stops on errors or it doesn't.
maybe it's better to stop the worker **before** it returns the error, so this state should become an `unreachable()`

Reply:
I've made sure the worker stops when it reaches a broken state